views/js/runner/plugin.js

```diff
@@ -0,0 +1,299 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2015 (original work) Open Assessment Technologies SA ;
+ */
+/**
+ *
+ * Runner plugin
+ *
+ * TODO usage example
+ *
+ * @author Sam <sam@taotesting.com>
+ * @author Bertrand Chevrier <bertrand@taotesting.com>
+ */
+define([
+    'lodash',
+    'core/promise'
+], function (_, Promise){
+    'use strict';
+
+    /**
+     * Meta factory for plugins. Let's you create a plugin definition.
+     *
+     * @param {Object} provider - the list of implemented methods
+     * @param {String} provider.name - the plugin name
+     * @param {Function} provider.init - the plugin initialization method
+     * @param {Function} [provider.render] - plugin render behaviorV
+     * @param {Function} [provider.finish] - plugin render behaviorV
+     * @param {Function} [provider.destroy] - plugin destroy behavior
+     * @param {Function} [provider.show] - plugin show behavior
+     * @param {Function} [provider.hide] - plugin hide behavior
+     * @param {Function} [provider.enable] - plugin enable behavior
+     * @param {Function} [provider.disable] - plugin disable behavior
+     * @param {Object} defaults - default configuration to be assigned
+     * @returns {Function} - the generated plugin factory
+     */
+    function pluginFactory(provider, defaults){
+        var pluginName;
+
+        if(!_.isPlainObject(provider) || !_.isString(provider.name) || _.isEmpty(provider.name) || !_.isFunction(provider.init)){
+            throw new TypeError('A plugin should be defined at least by a name property and an init method');
+        }
+
+        //TODO verify the name isn't already in use
+        pluginName = provider.name;
+
+        defaults = defaults || {};
+
+
+        /**
+         * The configured plugin factory
+         *
+         * @param {testRunner} runner - a test runner instance
+         * @param {areaBroker} areaBroker - an instance of an areaBrokee
+         * @param {Object} [config] - plugin configuration
+         * @returns {plugin} the plugin instance
+         */
+        return function instanciatePlugin(runner, areaBroker, config){
+            var plugin;
+
+            var states = {};
+
+            /**
+             * Delegate a function call to the provider
+             *
+             * @param {String} fnName - the function name
+             * @param {...} args - additional args are given to the provider
+             * @returns {*} up to the provider
+             */
+            function delegate(fnName){
+                var args = [].slice.call(arguments, 1);
+                return new Promise(function(resolve){
+                    if(!_.isFunction(provider[fnName])){
+                        resolve();
+                    }
+                    resolve(provider[fnName].apply(plugin, args));
```

Comment:
Maybe I'm wrong, but I not sure the `resolve` function breaks the execution flow, so an exception can be thrown if the provider method is not a function.
In the other hand, if no exception is thrown the `resolve` function may be called twice when the provider method is not defined or is not a function.